### ISSUE TITLE
Presence of required signatures not verified during contract verification

### ISSUE THREAD

**Issue by mvdbos** [corda/corda#5312]
This might be considered nitpicking, but in a space where everything is about trust, I consider this an issue: the fact that a tx is actually signed by the required signers is not confirmed within the contract verification itself.

TL;DR

* Why are signatures not passed to `Contract.verify()` as part of the LedgerTransaction? This seems intentional.
* `CommandWithParties.signers` is not the right name. It should be `CommandWithParties.requiredSigners`. Within the contract, we can't be sure that they actually signed.

Long version:

The signatures are not part of the `LedgerTransaction` passed to the contract's `verify` method. The signatures *are* checked by the ValidatingNotaryFlow, but how strong do we consider this guarantee? Why not simply also pass the signatures to the contract and allow contracts to do their own verification? IMO that would be stronger, and less dependent on the flows around the contract, even if we consider them part of the 'platform'.

Judging by the following comment on `LedgerTransaction.verify()` ([here](https://github.com/corda/corda/blob/b4e96778bfb157ac03ca97a2e9c9e76ac9bdd994/core/src/main/kotlin/net/corda/core/transactions/LedgerTransaction.kt#L124)) this is an intentional situation:

```
Verifies this transaction and runs contract code. At this stage it is assumed that signatures have already been verified.
```

I tried to come up with a reason *not* to pass the signatures to the contract as part of the LedgerTransaction, but couldn't. Perhaps someone form the Corda team can explain it?

This setup means that a very common contract constraint like the ones below have less meaning than people might think. If we can't be 100% sure that the signatures were checked outside the contract verification before, the only thing we are really checking here, is *that a certain pubkey is part of the list of required signers*. We don't actually know for sure within the contract that it was actually signed by them.

From CommercialPaper contract:

```kotlin
"the transaction is signed by the owner of the CP" using (input.owner.owningKey in command.signers)
```

From Cash contract:

```kotlin
    // Now check the digital signatures on the move command. Every input has an owning public key, and we must
    // see a signature from each of those keys. The actual signatures have been verified against the transaction
    // data by the platform before execution.
    val owningPubKeys = inputs.map { it.owner.owningKey }.toSet()
    val commands = commands.groupCommands<T>()
   [...]
    val keysThatSigned = commands.values.first()
    requireThat {
        "the owning keys are a subset of the signing keys" using keysThatSigned.containsAll(owningPubKeys)
    }
```

So even if we believe that the guarantees around signature checking are strong enough, I believe that `CommandWithParties.signers` is not the right name. It should be `requiredSigners`, so that it is clear for everyone that we are not actually checking the presence of the signature, but only the presence of a signature requirement.

**Comment by mvdbos** [comment]
Well, got some good and very obvious reasoning for the first question: at various points when the contract is verified, for example during signature collection between counterparties, not all required signatures are yet present. That makes it difficult to have the contract verify them. The only time when we know they should all be present is when notarising. I guess we'll just have to trust that the notary checks the sigs.

That leaves the second point: the in my opinion slightly misleading naming of CommandWithParties.signers.

**Comment by lankydan** [comment]
`CommandWithParties.signers` is unlikely to be changed since it is public api. We could choose to deprecate it and create another property with the same name that is backed by the original property. That being said, a comment is probably good enough here to clarify what it does.

**Comment by mvdbos** [comment]
Indeed, I understand changing public APIs is not on the cards. A JavaDoc comment is probably a good compromise. Shall I make a PR?

**Comment by mnesbit** [comment]
This would potentially cause a circularity issue. You need to verify locally before signing the transaction as one should not give away a signature on an invalid object. As a result the checks occur elsewhere during processing and during storage.

**Comment by mikehearn** [comment]
PRs to clarify welcome, but as you've seen, it is intentional. Here's a bit more background.

The validity of a transaction depends on many things. The contract verify method is just one of them. Here are some others:

* The constructors of deserialised objects are also considered to be (logically) a part of the smart contract. If an exception is thrown i.e. assertion fires during state construction, that means the transaction isn't valid. Thus c'tors define _internal validity_ (does the object's data make sense), and the verify method defines _transitional validity_ (does the change between two object graphs make sense).
* The inputs must exist and be themselves valid. User logic doesn't check this, the platform does.
  * In future SGX attestations or ZKPs may be used to show input validity instead of resolution.
* The signatures must be valid. The platform checks this, because that way we can check them in parallel, and also migrate to new signature schemes without needing every app to be upgraded.
* For some kinds of verification, the set of signatures must be complete, i.e. match the list of signers on the commands. But as you've already learned, you often want to check validity *before* signing a transaction, so there'd be circularity if we always insisted on this.
* There are transaction types that don't run contract logic, like notary change transactions.
* The contract constraints must be satisfied by the attachments. The platform checks this.
* The attachments must not overlap. The platform checks this.
* The transaction must not be too large. The platform checks this.

etc etc. There are many consensus rules involved and the verify method passing is only one of them.

Where are these rules checked? In system provided flows and logic. It must be so, because that's what defines the behaviour of your node. You could run a different node that doesn't do these checks, but the next node along the chain of custody will check them so you'd better be sure you won't get stuck.

**Comment by mvdbos** [comment]
All, thanks for taking the time to reply thoroughly. I understand how it is unavoidable that we give some of the transaction validity checking responsibility to the platform (system provided flows). This is for multiple reasons: not just logic (circularity) but also performance and flexibility. 

And while I agree that the next node along the chain of custody will challenge any wrongly verified transaction, I can dream of a future where finality is guaranteed to be achieved only for valid transactions. :-)

**Comment by mikehearn** [comment]
One thing we want to do is generalise the verifying notary concept to "signatures of validity" (primarily for SGX). So if you're willing to outsource transaction checking you can just check a signature. Might be useful for smaller programs or devices.
